rllib/evaluation/collectors/simple_list_collector.py

```diff
@@ -191,7 +215,101 @@ def add_action_reward_next_obs(self, values: Dict[str, TensorType]) -> None:
+        np_data = {}
+        for view_col, view_req in self.view_requirements.items():
+            # Create the batch of data from the different buffers.
+            data_col = view_req.data_col or view_col
+
+            # if this view is not for inference, skip it.
+            if not view_req.used_for_compute_actions:
+                continue
+            else:
+                if np.any(view_req.shift_arr > 0):
+                    raise ValueError(
+                        f"During inference the agent can only use past observations to "
+                        f"respect causality. However, view_col = {view_col} seems to "
+                        f"depend on future indices {view_req.shift_arr}, while the "
+                        f"used_for_compute_actions flag is set to True. Please fix the "
+                        f"discrepancy. Hint: If you are using a custom model make sure "
+                        f"the view_requirements are initialized properly and is point "
+                        f"only refering to past timesteps during inference."
+                    )
+
+            # Some columns don't exist yet
+            # (get created during postprocessing or depend on state_out).
+            if data_col not in self.buffers:
+                try:
+                    space = self.view_requirements[data_col].space
+                except KeyError:
+                    space = view_req.space
+
+                # special treatment for state_out_<i>
+                # add them to the buffer in case they don't exist yet
+                if data_col.startswith("state_out"):
+                    if not self.is_policy_recurrent:
+                        raise ValueError(
+                            f"{data_col} is not available, because the given policy is"
+                            f"not recurrent according to the input model_inital_states."
+                            f"Have you forgotten to return non-empty lists in"
+                            f"policy.get_initial_states()?"
+                        )
+                    state_ind = int(data_col.split("_")[-1])
+                    self._build_buffers({data_col: self.intial_states[state_ind]})
+                else:
+                    if isinstance(space, Space):
+                        fill_value = get_dummy_batch_for_space(
+                            space,
+                            batch_size=1,
+                        )
+                    else:
+                        fill_value = space
+
+                    self._build_buffers({data_col: fill_value})
+
+            # Keep an np-array cache so we don't have to regenerate the
+            # np-array for different view_cols using to the same data_col.
+            if data_col not in np_data:
+                np_data[data_col] = [
+                    _to_float_np_array(d) for d in self.buffers[data_col]
+                ]
+
+            data = []
+            for d in np_data[data_col]:
```

Comment:
I am curious, can any of these be shared with build()?
I kind of feel like these should be same logics? just with different set of VRs?

Reply:
ok shared most of the logic now. But the rest that are not shared have subtle differences that cannot be shared so easily.

Reply:
ok cool. thanks.